### PR TITLE
Sort courses on home page by date ascending

### LINKED ISSUE #365: Sort courses on home page by date ascending
As a user, I'd like to see courses on the home page with the soonest (future) start date first. 

If a course has no start date in the future, it should be at the bottom of the page. 

#### Acceptance Criteria:

- [x] #367
- [x] Courses with no future start date, should be at the bottom of the page

#### Related Issues

- the same sorting rules should apply to the catalog page, when we implement it

### PR DESCRIPTION
**Pull request by umarmughal824**
#### Pre-Flight checklist

- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested

- [x] Migrations
   - [x] Migration is backwards-compatible with current production code

#### What are the relevant tickets?
fixes #365

#### What's this PR do?
Sort courses on the home page by date ascending
- The course with the soonest, future start date should appear at the top of the page. #367
- Courses with no future start date, should be at the bottom of the page

#### How should this be manually tested?
Just add 5 - 10 feature courses (some with course runs in the past and rest with future nearest to the current date course runs) on the home page in cms and see that courses order

#### Screenshots (if appropriate)
**NOW**

Screenshot 2022-01-12 at 18 04 07

Screenshot 2022-01-14 at 12 49 53

**BEFORE**

Screenshot 2022-01-12 at 14 19 45

### CODE REVIEW

**Comment by asadiqbal08** [comment]
Here I have one question, If the course start date is passed then is it required behaviour to show it on last of the page as shown in your screenshots ?

File: cms/api_test.py [hunk 1]
```diff
@@ -134,3 +134,19 @@ def test_ensure_product_index(mocker):
...
+    run = existing_course_page.product.first_unexpired_run
+    assert featured_products == [
+        {
+            "title": existing_course_page.title,
+            "description": existing_course_page.description,
+            "feature_image": existing_course_page.feature_image,
+            "start_date": run.start_date if run is not None else None,
+            "url_path": existing_course_page.get_url(),
+        }
+    ]
```
Comment (asadiqbal08) [inline cms/api_test.py:152]: @umarmughal824 can you add a test where we have more than 1 feature products and we assert the order of items on basis of soonest future date

**Review by asadiqbal08 — CHANGES_REQUESTED** [review]

**Comment by asadiqbal08** [comment]
Just for information, I believe, so the logic in the PR now will eliminate the feature of wagtail of ordering feature products over HomePage. 

Screenshot 2022-01-13 at 2 56 35 PM

File: cms/api_test.py (hunk 1 above)
Comment (umarmughal824) [reply cms/api_test.py:152]: Let's wait for @pdpinch review, then will address the entire feedback altogether.

File: cms/models.py [hunk 2]
```diff
@@ -92,19 +93,34 @@ def _get_child_page_of_type(self, cls):
...
+                    "start_date": run.start_date if run is not None else None,
+                    "url_path": product_page.get_url(),
+                }
+                if run and run.start_date and run.start_date < now_in_utc():
+                    past_data.append(run_data)
+                else:
+                    page_data.append(run_data)
+
+        # sort future course run in ascending order
+        page_data = sorted(
```
Comment (pdpinch) [inline cms/models.py:114]: Nit: wondering if this variable name should be changed to reflect that it contains future concourses. For example, `future_data` similar to `past_data`

File: cms/api_test.py (hunk 1 above)
Comment (pdpinch) [reply cms/api_test.py:152]: What are "featured products" ?

**Review by pdpinch — APPROVED** [review]
This is looking good to me. I left a couple of small notes, but you don't have to change anything. 

> the logic in the PR now will eliminate the feature of wagtail of ordering feature products (moving up / moving down) over HomePage.

Thanks for bringing this up. I discussed with the marketing team, and they prefer to have this automated sort instead of a manual sort. 

> If the course start date is passed then is it required behaviour to show it on last of the page as shown in your screenshots ?

That's correct. We still want courses with start dates in the passed to be displayed, so that users know that we plan to run them again, but nothing is scheduled yet. If the course won't be run again, we will unpublish the page from wagtatil.

File: cms/api_test.py (hunk 1 above)
Comment (asadiqbal08) [reply cms/api_test.py:152]: referring the term when adding product pages in `HomePage`

File: cms/api_test.py [hunk 3]
```diff
@@ -134,3 +134,57 @@ def test_ensure_product_index(mocker):
     # Make sure the function is idempotent
     ensure_product_index()
     assert list(course_index_children_qset.all()) == [existing_course_page.page_ptr]
+
+
+@pytest.mark.django_db
+def atest_home_page_featured_products(mocker):
```
Comment (asadiqbal08) [inline cms/api_test.py:140]: nit: typo  `atest`

**Review by asadiqbal08 — APPROVED** [review]
just a nit, 👍

**Pull request merged**